Helper/NotifyEventTrait.php

```diff
@@ -33,16 +38,81 @@ public function setEventDispatcher(EventDispatcherInterface $eventDispatcher)
     }
 
     /**
-     * @param string $origin origin (producer or consumer)
-     *
-     * @return void
+     * @param string $origin
+     * @param string $key
      */
-    public function notifyEvent(string $origin)
+    public function prepareEvent(string $origin, string $key)
     {
         if ($this->eventDispatcher) {
+            $event = new KafkaEvent($origin);
+            $event->setExecutionStart();
+
+            $this->events[$key] = $event;
+        }
+    }
+
+    /**
+     * @param string $key
+     */
+    public function notifyEvent(string $key)
+    {
+        $event = $this->getEvent($key);
+        if ($this->eventDispatcher && $event) {
+            $event->setExecutionStop();
+
             $this
                 ->eventDispatcher
-                ->dispatch('kafka.event', new KafkaEvent($origin));
+                ->dispatch(KafkaEvent::EVENT_NAME, $event);
+        }
+    }
+
+    /**
+     * @param string $key
+     * @param int    $errorCode
+     */
+    public function notifyResponseErrorEvent(string $key, int $errorCode)
+    {
+        $event = $this->getEvent($key);
+        if ($this->eventDispatcher && $event) {
+            $event->setExecutionStop();
+            $event->setErrorCode($errorCode);
+            $event->setReason(rd_kafka_err2str($errorCode));
+
+            $this
+                ->eventDispatcher
+                ->dispatch(KafkaEvent::EVENT_ERROR_NAME, $event);
+        }
+    }
+
+    /**
+     * @param string $origin
+     * @param int    $errorCode
+     * @param string $reason
+     */
+    public function notifyErrorEvent(string $origin, int $errorCode, string $reason)
```

Comment:
Shouldn't we call it "notifyErrorFromProducerEvent" cause it deals only with producers?